src/sections/layer_and_mask_information_section/mod.rs

```diff
@@ -95,33 +96,35 @@ impl LayerAndMaskInformationSection {
         // channel as transparency data for the merged result.. So add a new test with a transparent
         // PSD and make sure that we're handling this case properly.
         let layer_count: u16 = layer_count.abs() as u16;
+        let mut layers = KeyIDContainer::with_capacity(layer_count as usize);
+        let mut groups = KeyIDContainer::new();
 
         let mut layer_records = vec![];
-
         // Read each layer record
         for _layer_num in 0..layer_count {
             layer_records.push(read_layer_record(&mut cursor)?);
         }
 
-        let mut sub_layers = PsdLayerGroup::new();
-
-        let mut parse_group_now = false;
+        let mut range_stack = vec![];
+        let mut nested_level = 0;
         // Read each layer's channel image data
         for (idx, layer_record) in layer_records.into_iter().enumerate() {
+            let group_id = if nested_level > 0 {
+                Some(groups.len() as u32)
+            } else {
+                None
+            };
+
             let mut psd_layer = PsdLayer::new(
-                layer_record.name.clone(),
-                layer_record.top,
-                layer_record.left,
-                layer_record.bottom,
-                layer_record.right,
+                layer_record.clone(),
                 psd_width,
                 psd_height,
-                None,
+                group_id,
             );
 
             let scanlines = layer_record.height() as usize;
 
-            for (channel_kind, channel_length) in layer_record.channel_data_lengths {
+            for (channel_kind, channel_length) in layer_record.clone().channel_data_lengths {
```

Comment:
Why do we need the clone? 

Seems like channel_kind and channel_length should both implement `Copy`? Or am I missing something?

So this could be `layer_record.channel_data_lengths.iter()`

And if we need to own them we could `*channel_kind` or `*channel_length` wherever necessary